src/api/db/models/utils.ts

```diff
@@ -0,0 +1,705 @@
+      // has an object that is locked,
+      // and its first validated label is included in labels filter
+      {
+        objects: {
+          $elemMatch: {
+            locked: true,
+            'firstValidLabel.labelId': { $in: labels },
+          },
+        },
+      },
+
+      // has an object is not locked, but it has label that is
+      // not-invalidated and included in filters
+      {
+        objects: {
+          $elemMatch: {
+            locked: false,
+            labels: {
+              $elemMatch: {
+                'validation.validated': { $not: { $eq: false } },
+                labelId: { $in: labels },
+              },
+            },
+          },
+        },
+      },
+    ],
+  };
+
+  // if labels includes "none", also return images with no objects
+  if (labels.includes('none')) {
+    const noObjectsFilter = {
+      $or: [
+        // return images w/ no objects,
+        { objects: { $size: 0 } },
+        // or images in which all labels of all objects have been invalidated
+        {
+          objects: {
+            $not: {
+              $elemMatch: {
+                labels: {
+                  $elemMatch: {
+                    $or: [{ validation: null }, { 'validation.validated': true }],
+                  },
+                },
+              },
+            },
+          },
+        },
+      ],
+    };
+    labelsFilter.$or.push(noObjectsFilter);
+  }
+
+  pipeline.push({ $match: labelsFilter });
+
+  return pipeline;
+}
+
+export function buildPipeline(
+  {
+    cameras,
+    deployments,
+    createdStart,
+    createdEnd,
+    addedStart,
+    addedEnd,
+    labels,
+    reviewed,
+    notReviewed,
+    custom,
+  }: gql.Filters,
+  projectId?: string,
+): PipelineStage[] {
+  const pipeline: PipelineStage[] = [];
+
+  // match current project
+  if (projectId) {
+    pipeline.push({ $match: { projectId: projectId } });
+  }
+
+  // match cameras filter
+  if (cameras) {
+    pipeline.push({ $match: { cameraId: { $in: cameras } } });
+  }
+
+  // match deployments filter
+  if (deployments) {
+    // cast string id to ObjectId
+    const deploymentIds = deployments.map((depString) => new mongoose.Types.ObjectId(depString));
+    pipeline.push({ $match: { deploymentId: { $in: deploymentIds } } });
+  }
+
+  // match date created filter
+  if (createdStart || createdEnd) {
+    pipeline.push({
+      $match: {
+        dateTimeOriginal: {
+          ...(createdStart && { $gte: createdStart.toJSDate() }),
+          ...(createdEnd && { $lt: createdEnd.toJSDate() }),
+        },
+      },
+    });
+  }
+
+  // match date added filter
+  if (addedStart || addedEnd) {
+    console.log('adding date added filter');
+    pipeline.push({
+      $match: {
+        dateAdded: {
+          ...(addedStart && { $gte: addedStart.toJSDate() }),
+          ...(addedEnd && { $lt: addedEnd.toJSDate() }),
+        },
+      },
+    });
+  }
+
+  // match notReivewed filter
+  if (reviewed === false) {
```

Comment:
Looks like this code is out of sync with models/utils.js. Please see: https://github.com/tnc-ca-geo/animl-api/blob/03b284db07a2e2753f9e122f6bcf8c7d0bf82f5a/src/api/db/models/utils.js#L147

Reply:
@alukach Also I performed a merge with main instead of a rebase just to resolve a merge conflict with main instead of rebasing just because I was less familiar with the changes made. Feel free to revert the merge and do a proper rebase

Reply:
Good catch.  That's a bit concerning, my `utils.ts` must have been created from `utils.js` before #172 was merged to `main`?  Not sure how that would have happened...

Reply:
Should be fixed in [6d375a0](https://github.com/tnc-ca-geo/animl-api/pull/218/commits/6d375a07a010a444cf0914379fb68317c89124b3)

Reply:
@jue-henry can you confirm that @alukach pulled in the latest `utils.js` code into the new `utils.ts`?

Reply:
yup looks good now